sample-notificator-app/src/main/groovy/life/qbic/samplenotificator/datasource/notification/create/CreateNotificationDbConnector.groovy

```diff
@@ -0,0 +1,139 @@
+
+import java.sql.Connection
+import java.sql.PreparedStatement
+import java.sql.Timestamp
+import java.time.Instant
+import java.time.LocalDate
+import java.time.ZoneId
+
+/**
+ * <h1>A database connector to retrieve information for sending notifications</h1>
+ *
+ * <p>This connector reads the notification queue table and finds the associated people that subscribed for a notification</p>
+ *
+ * @since 1.0.0
+ *
+*/
+@Log4j2
+class CreateNotificationDbConnector implements CreateNotificationDataSource{
+    private ConnectionProvider connectionProvider
+
+    CreateNotificationDbConnector(ConnectionProvider connectionProvider){
+        this.connectionProvider = connectionProvider
+    }
+
+
+    @Override
+    List<Subscriber> getSubscribersForNotificationsAt(LocalDate today) {
+        List<Subscriber> subscriberList = []
+        try{
+            //1. get todays notifications
+            Map<String, Status> sampleToStatus = getNotificationsForDay(today)
+            // retrieve the project code
+            Map<Integer,List<String>> subscriberIdsToSamples = getSubscriberIdForSamples(sampleToStatus)
+            //2. get the subscribers for the subscriptions
+            subscriberIdsToSamples.each { subscriberMap ->
+                Map<String,Status> allSamplesToStatus = sampleToStatus.findAll {it.key in subscriberMap.value}
+                subscriberList << getSubscriber(subscriberMap.key,allSamplesToStatus)
+            }
+        }catch(Exception e){
+            log.error e.message
+            log.error(e.stackTrace.join("\n"))
+
+            throw new DatabaseQueryException(e.message)
+        }
+
+        return subscriberList
+    }
+
+    private Map<String, Status> getNotificationsForDay(LocalDate day){
+        Instant startOfDay = day.atStartOfDay().atZone(ZoneId.systemDefault()).toInstant()
+        Instant endOfDay = day.plusDays(1).atStartOfDay().minusNanos(1).atZone(ZoneId.systemDefault()).toInstant()
+        String sqlQuery = SELECT_NOTIFICATIONS + " WHERE arrival_time BETWEEN ? AND ?"
+        Map<String, Status> foundNotifications = new HashMap<>()
+
+        Connection connection = connectionProvider.connect()
+
+        connection.withCloseable {
+            PreparedStatement preparedStatement = it.prepareStatement(sqlQuery)
+            preparedStatement.setTimestamp(1, Timestamp.from(startOfDay))
+            preparedStatement.setTimestamp(2, Timestamp.from(endOfDay))
+            preparedStatement.execute()
+
+            def resultSet = preparedStatement.getResultSet()
+            while (resultSet.next()) {
+                String sampleCode = resultSet.getString("sample_code")
+                Status status = Status.valueOf(resultSet.getString("sample_status"))
+                foundNotifications.put(sampleCode,status)
+            }
+        }
+        return foundNotifications
+    }
+
+    private Map<Integer,List<String>> getSubscriberIdForSamples(Map<String,Status> sampleToStatus){
```

Comment:
What is the intention of this method? I do not understand it I think. Is this a join in groovy?

Reply:
I think this is something that can be tackled by someone working on the use case? I think its too much to also work on that. I would opt for taking it as is and rework it